plugins/plugin-docker/che-plugin-docker-client/src/main/java/org/eclipse/che/plugin/docker/client/OpenShiftConnector.java

```diff
@@ -585,14 +594,90 @@ private boolean isTerminalAgentInjected(final Set<String> exposedPorts) {
     private boolean isDevMachine(final Set<String> exposedPorts) {
         return exposedPorts.contains(CHE_WORKSPACE_AGENT_PORT + "/tcp");
     }
-    
+
     /**
      * Che workspace id is used as OpenShift service / deployment config name
      * and must match the regex [a-z]([-a-z0-9]*[a-z0-9]) e.g. "q5iuhkwjvw1w9emg"
-     * 
+     *
      * @return randomly generated workspace id
      */
     private String generateWorkspaceID() {
         return RandomStringUtils.random(16, true, true).toLowerCase();
     }
-}
\ No newline at end of file
+
+    /**
+     * Converts a map of labels to match Kubernetes label requirements. Names are limited
+     * to alphanumeric characters, {@code '.'}, {@code '_'} and {@code '-'}, and must start and end
+     * with an alphanumeric character, i.e. they must match the regex
+     * {@code ([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]}
+     *
+     * <p>Note that entry keys should begin with {@link OpenShiftConnector#CHE_SERVER_LABEL_PREFIX} and
+     * entries should not contain {@code '.'} or {@code '_'} before conversion;
+     * otherwise label will not be converted and included in output.
+     *
+     * @param labels Map of labels to convert
+     * @return Map of labels converted to DNS Names
+     * @see OpenShiftConnector#convertKubernetesNamesToLabels(Map)
+     */
+    protected Map<String, String> convertLabelsToKubernetesNames(Map<String, String> labels) {
+        Map<String, String> names = new HashMap<>();
+        for (Map.Entry<String, String> entry : labels.entrySet()) {
+            String key = entry.getKey();
+            String value = entry.getValue();
+
+            if (value == null) {
+                continue;
+            }
+            // Check for potential problems with conversion
+            if (!key.startsWith(CHE_OPENSHIFT_RESOURCES_PREFIX)) {
+                LOG.warn("Expected CreateContainerParams label key " + entry.getKey() +
+                         " to start with " + CHE_SERVER_LABEL_PREFIX);
+            } else if (key.contains(".") || key.contains("_") || value.contains("_")) {
+                LOG.error(String.format("Cannot convert label %s to DNS Name: contains '-' or '.'", entry.toString()));
+                continue;
+            }
+
+            // Convert keys: e.g. "che:server:4401/tcp:ref" -> "che.server.4401-tcp.ref"
+            key = key.replaceAll(":", ".").replaceAll("/", "_");
+            // Convert values: e.g. "/api" -> ".api" -- note values may include '-'
```

Comment:
Values with `':'` will currently be skipped. This method is written for the current labels produced by Che -- values are protocol (e.g. http), ref (e.g. tomcat8-debug), and path (e.g. /api). I've written it to deal with those cases specifically, since if this changes (e.g. labels start to include other special characters), chances are more in-depth changes will be necessary.

This is kind of tricky -- the conversion methods are written for the specific structure of Che labels. Values don't currently seem to contain `':'`, so while the method could be written to handle that case, there is no reason to assume that that is the special character that *will* be used. Basically, I don't think we should use up one of the allowed special characters for an unnecessary case. 

Example: We replace `':'` and `'/'` now, but the replacement of `':'` is not necessary. Later, Che adds labels that contain `'*'` and `'#'`. Since we are already using `'.'` and `'_'` as replacement chars, it looks like we can't support the new labels. The person fixing the issue would have to look at labels to see that values don't contain `':'`. 

Let me know if this makes sense; I'm not terribly happy with how fragile these methods are, so I'm open to suggestions.